blog/_posts/2016-12-xx-moredots.md

```diff
@@ -0,0 +1,468 @@
+will also be applied elementwise to arrays, and either
+write it specially or define a corresponding elementwise method.
+
+(Our function `f` accepts any `x` type, and in Matlab or R there is no distinction between
+a scalar and a 1-element array.  However, even if a function *accepts* an array argument `x`,
+that doesn't mean it will *work* elementwise
+for an array unless you write the function with that in mind.)
+
+For library functions like `sqrt`, this means that the library authors
+have to guess at which functions should have vectorized methods, and users
+have to guess at what vaguely defined subset of library functions work
+for vectors.
+
+One possible solution is to vectorize *every function automatically*.   The
+language [Chapel](https://en.wikipedia.org/wiki/Chapel_%28programming_language%29)
+does this: every function `f(x...)` implicitly
+defines a function `f(x::Array...)` that evaluates `map(f, x...)`
+[(Chamberlain et al, 2011)](http://pgas11.rice.edu/papers/ChamberlainEtAl-Chapel-Iterators-PGAS11.pdf).
+This could be implemented in Julia as well via
+function-call overloading [(Bezanson, 2015: chapter 4)](https://github.com/JeffBezanson/phdthesis/blob/master/main.pdf),
+but we chose to go in a different direction.
+
+Instead, starting in Julia 0.5, *any* function `f(x)` can be applied elementwise
+to an array `X` with the ["dot call" syntax `f.(X)`](http://docs.julialang.org/en/stable/manual/functions/#dot-syntax-for-vectorizing-functions).
+Thus, the *caller* decides which functions to vectorize.  In Julia 0.6,
+"traditionally" vectorized library functions like `sqrt(X)` are [deprecated](https://github.com/JuliaLang/julia/pull/17302) in
+favor of `sqrt.(X)`, and dot operators
+like `x .+ y` are [now equivalent](https://github.com/JuliaLang/julia/pull/17623) to
+dot calls `(+).(x,y)`.   (Unlike Chapel's implicit vectorization, Julia's
+`f.(x...)` syntax corresponds to `broadcast(f, x...)` rather than `map`,
+allowing you to combine arrays of different shapes/dimensions.)
+From the standpoint of the programmer, this adds a certain amount of
+clarity because it indicates explicitly when an elementwise operation
+is occuring.  From the standpoint of the compiler, dot-call syntax
+enables the *syntactic loop fusion* optimization described in more detail
+below, which we think is an overwhelming advantage of this style.
+
+### Why vectorized code is fast
+
+In many dynamic languages for technical computing, vectorization is
+seen as a key (often *the* key) performance optimization.   It allows
+your code to take advantage of highly optimized (perhaps even parallelized)
+library routines for basic operations like `scalar*array` or `sqrt(array)`.
+Those functions, in turn, are usually implemented in a low-level language
+like C or Fortran.   Writing your own "devectorized" loops, in contrast,
+is too slow, unless you are willing to drop down to a low-level language
+yourself, because the semantics of those languages make it hard to
```

Comment:
"those languages" isn't obviously about the "dynamic languages for..." on first reading